In [2]:
import wikipedia
import numpy as np
import os
import json
import requests

In [3]:
code2lang = wikipedia.languages()
langs = ['ru', 'uk', 'en', 'nl', 'de', 'bg']

In [95]:
def get_texts_for_lang(lang, n=100): # функция для скачивания статей из википедии
    wikipedia.set_lang(lang)
    wiki_content = []
    pages = wikipedia.random(n)
    
    for page_name in pages:
        try:
            page = wikipedia.page(page_name)
        
        except Exception as e:
            print('Skipping page {}'.format(page_name), e)
            continue

        wiki_content.append('{}\n{}'.format(page.title, page.content.replace('==', '')))

    return wiki_content

In [96]:
wiki_texts = {}
for lang in langs:
    try:
        wiki_texts[lang] = get_texts_for_lang(lang, 100)
    except Exception as e:
        print('ERROR ON - ', lang, e)
        continue
    print(lang, len(wiki_texts[lang]))

c:\users\maria tsareva\appdata\local\programs\python\python36-32\lib\site-packages\bs4\__init__.py:181: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html5lib"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 193 of the file c:\users\maria tsareva\appdata\local\programs\python\python36-32\lib\runpy.py. To get rid of this warning, change code that looks like this:

 BeautifulSoup(YOUR_MARKUP})

to this:

 BeautifulSoup(YOUR_MARKUP, "html5lib")

  markup_type=markup_type))


Skipping page Смирнов, Юрий "Смирнов, Юрий" may refer to: 
Смирнов, Юрий Александрович
Смирнов, Юрий Александрович
Смирнов, Юрий Александрович
Смирнов, Юрий Валентинович
Смирнов, Юрий Васильевич
Смирнов, Юрий Васильевич
Смирнов, Юрий Васильевич
Смирнов, Юрий Васильевич
Смирнов, Юрий Константинович
Смирнов, Юрий Михайлович
Смирнов, Юрий Михайлович
Смирнов, Юрий Михайлович
Смирнов, Юрий Николаевич
Смирнов, Юрий Николаевич
Смирнов, Юрий Николаевич
Смирнов, Юрий Николаевич
Смирнов, Юрий Петрович
Skipping page Мордовский государственный университет имени Н. П. Огарёва Page id "московский государственный университет имени н п огарева" does not match any pages. Try another id!
Skipping page Куангнинь (значения) "Куангнинь (значения)" may refer to: 
Куангнинь
Куангнинь
Куангнинь
Skipping page Голубиха "Голубиха" may refer to: 
Голубиха (приток Узолы)
Голубиха
Голубиха
Skipping page HMS Challenger "HMS Challenger" may refer to: 
HMS Challenger (1858)
HMS Challenger (1902)
HMS Challenger (1931)


Частотные списки

In [56]:
import codecs
import collections
import sys
import re
import operator

def tokenize(text):
    text = text.lower()
    text = re.sub(r'\s+', ' ', text, re.UNICODE)
    text = re.sub(r'[^\w\s]', '', text, re.UNICODE)
    return text.split(' ')


def freq(wiki_texts, lang):
    freqs = collections.defaultdict(lambda: 0)
    words = []
    corpus = wiki_texts[lang]
    for article in corpus:
        for word in tokenize(article.replace('\n', '').lower()):
            freqs[word] += 1

    for word in sorted(freqs, key=lambda w: freqs[w], reverse=True)[:50]:
        words.append(word)
#         print('{}\t{}'.format(freqs[word], word))
    return words

In [31]:
freq_ = {}
for lang in langs:
    freq_[lang] = freq(wiki_texts, lang)

In [78]:
def coun(freq_, text):
    text = text.lower()
    i_ = {}
    i = 0
    for key, item in freq_.items():
        for word in item:
#         print(item)
            if word in text.split():
                i += 1
        i_[key] = i
        i = 0
    print(i_)
    return max(i_.items(), key=operator.itemgetter(1))[0]

In [80]:
text = wiki_texts
print(text)
cc = coun(freq_, text)
cc

W. A. Mason House
The W. A. Mason House is an historic house at 87 Raymond Street in Cambridge, Massachusetts. The 2-1/2 story wood frame house was built in 1846 and extended to the rear in 1867, giving it a T shape. Its massing is Italianate as are its decorative window hoods, but its corner pilasters give it a Greek Revival character. The main entrance is accessed through a single-story porch occupying the front crook of the T. W. A. Mason, for whom it was built, was a city surveyor who was responsible for surveying and platting a significant amount of the city in the 19th century.
The house was listed on the National Register of Historic Places in 1983.


 See also 
National Register of Historic Places listings in Cambridge, Massachusetts


 References 
{'ru': 1, 'uk': 0, 'en': 19, 'nl': 5, 'de': 2, 'bg': 2}


'en'

In [74]:
from sklearn.metrics import classification_report, confusion_matrix

In [97]:
true_labels = []
predicted_labels = []

for lang in wiki_texts:
    for text in wiki_texts[lang]:
        true_labels.append(lang)
        predicted_labels.append(coun(freq_,text))

{'ru': 29, 'uk': 10, 'en': 3, 'nl': 3, 'de': 1, 'bg': 13}
{'ru': 7, 'uk': 2, 'en': 0, 'nl': 0, 'de': 0, 'bg': 3}
{'ru': 28, 'uk': 8, 'en': 0, 'nl': 0, 'de': 1, 'bg': 11}
{'ru': 18, 'uk': 7, 'en': 2, 'nl': 3, 'de': 1, 'bg': 10}
{'ru': 26, 'uk': 9, 'en': 3, 'nl': 2, 'de': 2, 'bg': 15}
{'ru': 26, 'uk': 8, 'en': 0, 'nl': 0, 'de': 1, 'bg': 11}
{'ru': 23, 'uk': 8, 'en': 0, 'nl': 0, 'de': 1, 'bg': 10}
{'ru': 6, 'uk': 2, 'en': 0, 'nl': 0, 'de': 0, 'bg': 3}
{'ru': 12, 'uk': 4, 'en': 0, 'nl': 0, 'de': 0, 'bg': 5}
{'ru': 38, 'uk': 10, 'en': 1, 'nl': 1, 'de': 0, 'bg': 12}
{'ru': 32, 'uk': 10, 'en': 7, 'nl': 3, 'de': 2, 'bg': 14}
{'ru': 7, 'uk': 4, 'en': 0, 'nl': 0, 'de': 0, 'bg': 3}
{'ru': 15, 'uk': 7, 'en': 0, 'nl': 0, 'de': 0, 'bg': 9}
{'ru': 26, 'uk': 7, 'en': 1, 'nl': 0, 'de': 1, 'bg': 7}
{'ru': 15, 'uk': 5, 'en': 2, 'nl': 1, 'de': 0, 'bg': 6}
{'ru': 7, 'uk': 4, 'en': 8, 'nl': 5, 'de': 4, 'bg': 5}
{'ru': 16, 'uk': 6, 'en': 0, 'nl': 1, 'de': 5, 'bg': 7}
{'ru': 12, 'uk': 3, 'en': 0, 'nl': 0, 'de

{'ru': 11, 'uk': 38, 'en': 0, 'nl': 0, 'de': 0, 'bg': 9}
{'ru': 9, 'uk': 19, 'en': 0, 'nl': 1, 'de': 0, 'bg': 6}
{'ru': 15, 'uk': 22, 'en': 3, 'nl': 3, 'de': 3, 'bg': 12}
{'ru': 6, 'uk': 13, 'en': 0, 'nl': 0, 'de': 0, 'bg': 3}
{'ru': 9, 'uk': 16, 'en': 0, 'nl': 0, 'de': 0, 'bg': 6}
{'ru': 11, 'uk': 22, 'en': 1, 'nl': 2, 'de': 0, 'bg': 8}
{'ru': 7, 'uk': 18, 'en': 0, 'nl': 1, 'de': 0, 'bg': 4}
{'ru': 6, 'uk': 22, 'en': 3, 'nl': 2, 'de': 1, 'bg': 6}
{'ru': 4, 'uk': 8, 'en': 0, 'nl': 0, 'de': 1, 'bg': 3}
{'ru': 5, 'uk': 17, 'en': 0, 'nl': 0, 'de': 0, 'bg': 3}
{'ru': 7, 'uk': 16, 'en': 1, 'nl': 0, 'de': 1, 'bg': 4}
{'ru': 6, 'uk': 12, 'en': 0, 'nl': 0, 'de': 0, 'bg': 4}
{'ru': 4, 'uk': 10, 'en': 5, 'nl': 4, 'de': 1, 'bg': 3}
{'ru': 5, 'uk': 9, 'en': 0, 'nl': 0, 'de': 1, 'bg': 3}
{'ru': 8, 'uk': 12, 'en': 0, 'nl': 1, 'de': 0, 'bg': 5}
{'ru': 2, 'uk': 2, 'en': 0, 'nl': 0, 'de': 0, 'bg': 1}
{'ru': 4, 'uk': 6, 'en': 0, 'nl': 1, 'de': 2, 'bg': 1}
{'ru': 4, 'uk': 16, 'en': 0, 'nl': 0, 'de': 0, '

{'ru': 2, 'uk': 0, 'en': 13, 'nl': 31, 'de': 9, 'bg': 4}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 15, 'de': 2, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 16, 'de': 2, 'bg': 0}
{'ru': 2, 'uk': 0, 'en': 4, 'nl': 18, 'de': 2, 'bg': 2}
{'ru': 2, 'uk': 0, 'en': 4, 'nl': 24, 'de': 5, 'bg': 1}
{'ru': 2, 'uk': 0, 'en': 4, 'nl': 29, 'de': 4, 'bg': 2}
{'ru': 1, 'uk': 1, 'en': 4, 'nl': 21, 'de': 4, 'bg': 2}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 10, 'de': 2, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 3, 'nl': 16, 'de': 4, 'bg': 1}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 20, 'de': 4, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 17, 'de': 2, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 16, 'de': 2, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 17, 'de': 2, 'bg': 0}
{'ru': 2, 'uk': 0, 'en': 8, 'nl': 20, 'de': 3, 'bg': 3}
{'ru': 1, 'uk': 0, 'en': 4, 'nl': 16, 'de': 2, 'bg': 1}
{'ru': 1, 'uk': 0, 'en': 4, 'nl': 21, 'de': 4, 'bg': 1}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 8, 'de': 2, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 5, 'de': 2, 'b

{'ru': 0, 'uk': 0, 'en': 3, 'nl': 6, 'de': 41, 'bg': 1}
{'ru': 1, 'uk': 0, 'en': 5, 'nl': 7, 'de': 26, 'bg': 2}
{'ru': 0, 'uk': 0, 'en': 3, 'nl': 6, 'de': 40, 'bg': 1}
{'ru': 0, 'uk': 0, 'en': 1, 'nl': 3, 'de': 30, 'bg': 0}
{'ru': 2, 'uk': 0, 'en': 7, 'nl': 7, 'de': 37, 'bg': 2}
{'ru': 1, 'uk': 0, 'en': 2, 'nl': 7, 'de': 34, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 0, 'nl': 1, 'de': 6, 'bg': 1}
{'ru': 0, 'uk': 0, 'en': 1, 'nl': 6, 'de': 20, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 1, 'nl': 4, 'de': 20, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 1, 'nl': 5, 'de': 20, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 4, 'nl': 6, 'de': 34, 'bg': 3}
{'ru': 0, 'uk': 0, 'en': 1, 'nl': 3, 'de': 5, 'bg': 0}
{'ru': 5, 'uk': 3, 'en': 2, 'nl': 1, 'de': 0, 'bg': 10}
{'ru': 2, 'uk': 1, 'en': 0, 'nl': 0, 'de': 0, 'bg': 4}
{'ru': 10, 'uk': 6, 'en': 0, 'nl': 0, 'de': 2, 'bg': 22}
{'ru': 7, 'uk': 4, 'en': 4, 'nl': 3, 'de': 1, 'bg': 13}
{'ru': 13, 'uk': 7, 'en': 1, 'nl': 1, 'de': 3, 'bg': 27}
{'ru': 4, 'uk': 2, 'en': 0, 'nl': 0, 'de': 1, 'bg

In [98]:
print(classification_report(true_labels, predicted_labels))

             precision    recall  f1-score   support

         bg       1.00      1.00      1.00        97
         de       1.00      0.99      0.99        90
         en       0.97      1.00      0.98        92
         nl       1.00      1.00      1.00        98
         ru       0.96      0.99      0.97        91
         uk       1.00      0.95      0.97        96

avg / total       0.99      0.99      0.99       564



Набор N-gramm

In [83]:
from itertools import islice, tee

def make_ngrams(text):
    N = 3 # задаем длину n-граммы
    ngrams = zip(*(islice(seq, index, None) for index, seq in enumerate(tee(text, N))))
    ngrams = [''.join(x) for x in ngrams]
    return ngrams

In [86]:
def freq_ns(wiki_texts, lang):
    freqs = collections.defaultdict(lambda: 0)
    ns = []
    corpus = wiki_texts[lang]
    for article in corpus:
        for ngram in make_ngrams(article.replace('\n', '').lower()):
            freqs[ngram] += 1

    for ngram in sorted(freqs, key=lambda n: freqs[n], reverse=True)[:50]:
        ns.append(ngram)
#         print('{}\t{}'.format(freqs[word], word))
    return ns

In [87]:
freq_n = {}
for lang in langs:
    freq_n[lang] = freq_ns(wiki_texts, lang)

In [90]:
def coun_ns(freq_n, text):
    text = make_ngrams(text)
    i_ = {}
    i = 0
    for key, item in freq_n.items():
        for ns in item:
#         print(item)
            if ns in text:
                i += 1
        i_[key] = i
        i = 0
    print(i_)
    return max(i_.items(), key=operator.itemgetter(1))[0]

In [91]:
text = wiki_texts['en'][0]
print(text)
nn = coun_ns(freq_n, text)
nn

W. A. Mason House
The W. A. Mason House is an historic house at 87 Raymond Street in Cambridge, Massachusetts. The 2-1/2 story wood frame house was built in 1846 and extended to the rear in 1867, giving it a T shape. Its massing is Italianate as are its decorative window hoods, but its corner pilasters give it a Greek Revival character. The main entrance is accessed through a single-story porch occupying the front crook of the T. W. A. Mason, for whom it was built, was a city surveyor who was responsible for surveying and platting a significant amount of the city in the 19th century.
The house was listed on the National Register of Historic Places in 1983.


 See also 
National Register of Historic Places listings in Cambridge, Massachusetts


 References 
{'ru': 1, 'uk': 0, 'en': 45, 'nl': 16, 'de': 18, 'bg': 0}


'en'

In [99]:
true_labels = []
predicted_labels = []

for lang in wiki_texts:
    for text in wiki_texts[lang]:
        true_labels.append(lang)
        predicted_labels.append(coun_ns(freq_n,text))

{'ru': 49, 'uk': 34, 'en': 7, 'nl': 3, 'de': 4, 'bg': 44}
{'ru': 30, 'uk': 14, 'en': 0, 'nl': 0, 'de': 0, 'bg': 18}
{'ru': 49, 'uk': 33, 'en': 0, 'nl': 0, 'de': 0, 'bg': 39}
{'ru': 39, 'uk': 20, 'en': 7, 'nl': 6, 'de': 12, 'bg': 39}
{'ru': 46, 'uk': 33, 'en': 9, 'nl': 2, 'de': 4, 'bg': 39}
{'ru': 48, 'uk': 30, 'en': 0, 'nl': 0, 'de': 0, 'bg': 39}
{'ru': 46, 'uk': 31, 'en': 0, 'nl': 0, 'de': 0, 'bg': 32}
{'ru': 32, 'uk': 13, 'en': 0, 'nl': 0, 'de': 0, 'bg': 19}
{'ru': 33, 'uk': 20, 'en': 0, 'nl': 0, 'de': 0, 'bg': 26}
{'ru': 49, 'uk': 37, 'en': 7, 'nl': 1, 'de': 0, 'bg': 45}
{'ru': 49, 'uk': 34, 'en': 31, 'nl': 13, 'de': 16, 'bg': 40}
{'ru': 24, 'uk': 11, 'en': 0, 'nl': 0, 'de': 0, 'bg': 17}
{'ru': 44, 'uk': 27, 'en': 0, 'nl': 1, 'de': 1, 'bg': 33}
{'ru': 49, 'uk': 32, 'en': 0, 'nl': 0, 'de': 0, 'bg': 37}
{'ru': 39, 'uk': 25, 'en': 2, 'nl': 1, 'de': 2, 'bg': 25}
{'ru': 24, 'uk': 11, 'en': 28, 'nl': 16, 'de': 20, 'bg': 15}
{'ru': 47, 'uk': 29, 'en': 7, 'nl': 7, 'de': 11, 'bg': 30}
{'ru':

{'ru': 39, 'uk': 50, 'en': 27, 'nl': 10, 'de': 19, 'bg': 39}
{'ru': 17, 'uk': 28, 'en': 0, 'nl': 0, 'de': 0, 'bg': 17}
{'ru': 23, 'uk': 35, 'en': 3, 'nl': 0, 'de': 1, 'bg': 18}
{'ru': 17, 'uk': 17, 'en': 0, 'nl': 0, 'de': 0, 'bg': 8}
{'ru': 4, 'uk': 7, 'en': 0, 'nl': 0, 'de': 0, 'bg': 4}
{'ru': 26, 'uk': 40, 'en': 0, 'nl': 0, 'de': 0, 'bg': 13}
{'ru': 14, 'uk': 19, 'en': 0, 'nl': 0, 'de': 0, 'bg': 16}
{'ru': 36, 'uk': 49, 'en': 0, 'nl': 0, 'de': 0, 'bg': 36}
{'ru': 33, 'uk': 41, 'en': 0, 'nl': 0, 'de': 0, 'bg': 31}
{'ru': 47, 'uk': 45, 'en': 31, 'nl': 17, 'de': 22, 'bg': 41}
{'ru': 17, 'uk': 31, 'en': 0, 'nl': 0, 'de': 1, 'bg': 17}
{'ru': 28, 'uk': 44, 'en': 14, 'nl': 9, 'de': 13, 'bg': 23}
{'ru': 35, 'uk': 47, 'en': 7, 'nl': 2, 'de': 2, 'bg': 30}
{'ru': 27, 'uk': 46, 'en': 8, 'nl': 3, 'de': 5, 'bg': 19}
{'ru': 31, 'uk': 47, 'en': 3, 'nl': 4, 'de': 3, 'bg': 26}
{'ru': 18, 'uk': 28, 'en': 0, 'nl': 0, 'de': 0, 'bg': 12}
{'ru': 24, 'uk': 40, 'en': 0, 'nl': 0, 'de': 0, 'bg': 19}
{'ru': 31,

{'ru': 1, 'uk': 0, 'en': 42, 'nl': 50, 'de': 47, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 5, 'nl': 32, 'de': 11, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 26, 'nl': 38, 'de': 28, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 7, 'nl': 36, 'de': 11, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 24, 'nl': 45, 'de': 32, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 20, 'nl': 43, 'de': 33, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 48, 'nl': 49, 'de': 42, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 20, 'nl': 37, 'de': 26, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 9, 'nl': 39, 'de': 15, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 27, 'nl': 35, 'de': 24, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 29, 'nl': 47, 'de': 37, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 33, 'nl': 48, 'de': 41, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 36, 'nl': 44, 'de': 36, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 10, 'nl': 31, 'de': 19, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 24, 'nl': 36, 'de': 31, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 22, 'nl': 40, 'de': 28, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 8, 'nl': 36, 'de': 12, 'bg': 0}
{'ru': 0, 'uk': 0,

{'ru': 1, 'uk': 0, 'en': 18, 'nl': 29, 'de': 38, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 36, 'nl': 37, 'de': 50, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 22, 'nl': 26, 'de': 43, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 22, 'nl': 26, 'de': 41, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 37, 'nl': 38, 'de': 50, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 40, 'nl': 39, 'de': 50, 'bg': 0}
{'ru': 2, 'uk': 0, 'en': 44, 'nl': 39, 'de': 50, 'bg': 1}
{'ru': 1, 'uk': 0, 'en': 31, 'nl': 30, 'de': 48, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 36, 'nl': 38, 'de': 50, 'bg': 0}
{'ru': 3, 'uk': 2, 'en': 33, 'nl': 32, 'de': 48, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 33, 'nl': 37, 'de': 50, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 29, 'nl': 33, 'de': 48, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 41, 'nl': 34, 'de': 50, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 30, 'nl': 35, 'de': 50, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 10, 'nl': 9, 'de': 17, 'bg': 0}
{'ru': 1, 'uk': 0, 'en': 23, 'nl': 29, 'de': 49, 'bg': 0}
{'ru': 0, 'uk': 0, 'en': 22, 'nl': 30, 'de': 49, 'bg': 0}
{'ru': 1, 'uk':

In [100]:
print(classification_report(true_labels, predicted_labels))

             precision    recall  f1-score   support

         bg       1.00      0.99      0.99        97
         de       1.00      1.00      1.00        90
         en       0.95      1.00      0.97        92
         nl       1.00      0.98      0.99        98
         ru       0.98      0.99      0.98        91
         uk       1.00      0.97      0.98        96

avg / total       0.99      0.99      0.99       564



Отчет

Я использовала оба метода для распознавания языка. Каждый из них дал на текстах, созданных дополнительно функцией рандомного парсинга тестов из википедии, 0.99. Если смотреть отдельно для языков, то можно заметить, что в некоторых случаях лучше отработал метод частотных слов (английский - 0.98 против 0.95), в других - метод н-грамм (русский - 0.96 против 0.98).